[WIP] Create endpoints and model for dealing with local snaps

justincan opened the pull request.

mvo5 commented:
Just a general question for now - it seems like this PR adds more code to the snappy daemon to handle listing of upgradable snaps etc. I wonder if this is the right direction. It seems (given that we have a snapd rest API) that ideally we move towards more "passthrough" code in the daemon. It looks like most of this is just passing data through which makes me wonder why we need an addtional layer of unmarshal/marshal etc. I guess the JS code needs some adjustments because the snappy daemon snapweb api and the snapd api are slightly different. But it seems they are close enough. Or is there a more fundamental reason to keep this additional layer between snapd<->snapweb JS (expect for bridging between network and the unix socket of course)?

justincan commented:
RIght now it's for use with upgrades, but I would like to see the model used for all situations that require interaction with local snaps - it could change a little, it has this form to ease integration pain with other contributors - but the semantics should remain.

An example of the importance of the distinction - with REST style semantics, if I wish to change an existing entity I could use PUT (or PATCH). Does it make sense to PUT/PATCH an entity on the store as a snapweb user? If at some point users can manage _their_ on store snaps, then yes, but otherwise no, this way has future proofing as well :)

It would probably be for the best if the snapd API had a stronger distinction as well, then other functionality may "fall out" in a natural way - such as the handling of history/rollback.

I completely agree that as much functionality as possible is presented and used through the snapd API. But there are always going to be edges that need to be managed through an adaptation layer. If we wholly expect this to happen at the js level, then we are walking the road of writing YAJF, rather I think we should use each of the pieces as is and make changes to the code that has no other-party responsibilities; snapweb. It is going to be a lot easier than expecting/spending effort to change various frameworks.

Hopefully I've addressed your concerns, but please let me know if there is more.

justincan commented:
@AlexandreAbreu could you review this?

dbarth commented:
With hindsight I am not so keen on having that either : it creates another undocumented API, whereas snapd's API is official. Plus, I feel the adaptation can be more easily done in JS.
Last, the REST orthodoxy is motivated by having less to hack in Backbone, but that may become less of a requirement if switching to React/Redux.

So overall, I would like to see another branch which does the job in JS/Backone, and re-stack the other PRs (Notifications) without this one.

justincan commented:
I don't think it is advisable or even wholly possible to pursue a .js only solution.

1) Obtaining "refreshable" (nee updatable) snaps through the client api requires bypassing the default find options and using a query parameter that is otherwise hidden in snapd.

2) Adding an accurate representation of local snaps allows a contextually correct API for updating apps - a task which this PR enables and is only possible with .go code changes.

dbarth commented:
Closing this branch. The concept has been reviewed but not considered worth pursuing to interface with snapd. So, we stay with the current API which works. Thanks for proposing this anyway.
